### PR TITLE
Scrabble breakfast

### PR DESCRIPTION
**Pull request by lilagrc**
Completed project to breakfast level. I used awesome print for testing, so you may need to require it. I'm a little uncertain if all my edge cases work, but all the specs I wrote are coming through as passing. I was absent the first day of this project, so I tried to crank this out in one day and I don't feel awesome about my code. However, I think with more time/brain power I would have been able to better test for edge cases, clean up the code/logic that I used, and maybe even take a gander at lunch!

### CODE REVIEW

**Comment by kariabancroft** [comment]
Hi @lilagrc it looks like your RSpec tests are missing from your commits. Could you add them?

**Comment by lilagrc** [comment]
Yes! I just pushed the rspec file. Do I need to do another PR?

File: lib/scrabble.rb [hunk 1]
```diff
@@ -1,4 +1,92 @@
 module Scrabble
+
+LETTER_VALUES = [
+  [1, "A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
```
Comment (kariabancroft) [inline lib/scrabble.rb]: Based on the way you have structured this - it might make sense to use a hash since it seems like the first element in the array index is your "key"

File: lib/scrabble.rb [hunk 2]
```diff
@@ -1,4 +1,92 @@
...
+  [8, "J", "X"],
+  [10, "Q", "Z"]
+]
+
+class Scrabble
   # require your gems and classes here
   # require_relative 'scrabble/whatevs'
+  require "awesome_print"
+
+def self.letter_score(letter)
```
Comment (kariabancroft) [inline lib/scrabble.rb]: All of your methods inside the Scrabble class should be tabbed in rather than inline with the class definition

File: lib/scrabble.rb [hunk 3]
```diff
@@ -1,4 +1,92 @@
...
+  LETTER_VALUES.each do |letter_value_array|
+      if letter_value_array.include?(letter.upcase)
+        value = letter_value_array[0]
```
Comment (kariabancroft) [inline lib/scrabble.rb]: Setting this to the `value` variable would make sense if you were going to reuse `value`, but i think you could just use `return letter_value_array[0]` here

File: lib/scrabble.rb [hunk 4]
```diff
@@ -1,4 +1,92 @@
...
+        return value
+      end
+  end
 end
+
+def self.score(word)
+  #&& word_array.empty? == false (to test empty input)
+  #splits word into array of letters
+  word_array = word.upcase.split(//)
+  letters = ("A".."Z")
```
Comment (kariabancroft) [inline lib/scrabble.rb]: This would be a place for a constant since the default set of letters will not change

File: lib/scrabble.rb [hunk 5]
```diff
@@ -1,4 +1,92 @@
...
+
+  word_array.each do |letter|
+    unless letters.include?(letter) and word_array.length <= 7
+      return "Error"
+    end
+  end
+
+  #runs each letter through letter_score methods and scores it
+  word_values = []
+  word_array.each do |letter|
```
Comment (kariabancroft) [inline lib/scrabble.rb]: This iterator is the same as the one on lines 33-37 which raises a red flag of inefficiency. You should be able to include the error logic in this iterator so you only have to go through each letter in the word once.

File: lib/scrabble.rb [hunk 6]
```diff
@@ -1,4 +1,92 @@
...
+    value = Scrabble.letter_score(letter)
+    word_values << value
+  end
+
+  #sums array of individual letter values
+  score = 0
+  word_values.each do |value|
+    score += value
+  end
```
Comment (kariabancroft) [inline lib/scrabble.rb]: This summation is good - but could also be cleaned up. You could actually just keep incrementing the scores in the above loop rather than looping through the values after they have been calculated.

File: lib/scrabble.rb [hunk 7]
```diff
@@ -1,4 +1,92 @@
...
+
+  return score
+end
+
+def self.highest_score_from(array_of_words)
+  value_array = []
+
+  #pair scores with words in a new array
+  array_of_words.each do |word, value|
```
Comment (kariabancroft) [inline lib/scrabble.rb]: I don't think you want to use the `value` variable in your loop block here since it doesn't exist yet in your `array_of_words`

File: lib/scrabble.rb [hunk 8]
```diff
@@ -1,4 +1,92 @@
...
+
+  #compares scores in value_array with max_word_array to check for ties in scores
+  tied_words = []
+  value_array.each do |array|
+    if max_word_array[0] == array[0]
+      tied_words << array[1]
+    end
+  end
+
+  smallest_word = tied_words.min_by { |word| word.length }
```
Comment (kariabancroft) [inline lib/scrabble.rb]: This is a great use of the `min_by` - could you rewrite what you have below using `max_by`?

File: spec/scrabble_spec.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
+require './lib/scrabble'
+
+describe "Score a letter" do
+  it "returns the value for a single letter" do
+  expect(Scrabble::Scrabble.letter_score("d")).to eq (2)
+  end
+end
+
+describe "Score a word in Scrabble" do
+  context "word should be a string of letters" do
```
Comment (kariabancroft) [inline spec/scrabble_spec.rb:10]: Yes - very nice use of the context to contain the error cases here

File: spec/scrabble_spec.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+end #describe end
+
+describe "Find highest score in an array" do
+  it "returns word in array with highest score" do
+    expect(Scrabble::Scrabble.highest_score_from(["boy", "jazz", "dog"])).to eq("jazz")
+  end
+
+  context "if two words have the same score:" do
+      it "returns word with fewer tiles" do
+       expect(Scrabble::Scrabble.highest_score_from(["ok", "dog", "win"])).to eq("ok")
```
Comment (kariabancroft) [inline spec/scrabble_spec.rb]: This test would be better if "ok" was not the first string in the list - since it could just be abiding by the first choice rule that you're testing for later.

**Comment by kariabancroft** [comment]
Great job! There are a few refactoring opportunities here to make your code more efficient. Definitely keep an eye out for situations where you use the same or similar iterators one after another. Your tests were well-written using describe and context to ensure appropriate groupings.

**Comment by lilagrc** [comment]
I added one more commit after having some time to refactor in class. I didn't change my array in letters into a hash due since we ran out of time, but if I were to refactor in the future, that would be on the top of my list!
